### PR TITLE
Graph embedding and graph backbones

### LINKED ISSUE #510: Replace current BaseGraphModel (from #73) with GRAPH_CLASSIFICATION_BACKBONES FlashRegistry

### LINKED ISSUE #511: Add a GraphEmbedder Task

### LINKED ISSUE #512: Add a GRAPH_EMBEDDING_BACKBONES FlashRegistry

### LINKED ISSUE #513: Add a graph_embedding example based on our other examples

### LINKED ISSUE #514: Add a Graph Embedding page to the docs based on our other docs pages

### PR DESCRIPTION
**Pull request by PabloAMC**
## What does this PR do?

Fixes #510
Fixes #511 
Fixes #512 
Fixes #513 
Fixes #514 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
hey @PabloAMC, mind rebasing on master ?

**Comment by PabloAMC** [comment]
> hey @PabloAMC, mind rebasing on master ?

@tchaton I have checked and github says that the branch is up to date with master. Do you mean upstream/master ?

**Review by ethanwharris — COMMENTED** [review]
@PabloAMC This looks awesome! Do we have any estimated timeline for when the models will be added to torch-geometric? We could maybe merge this with just the GCN that is there now if there is a release. What are your thoughts?

**Comment by PabloAMC** [comment]
Hi @ethanwharris ! Actually, I got the basic model PR merged! (after some refactoring).
https://github.com/rusty1s/pytorch_geometric/pull/2869
 So now GCN, GIN, GAT and GraphSage are basic models that we should be able to call (or perhaps in the next pytorch geometric version). I have to make the final tweaks to this branch though to get only those models
 
 Then there is also the other PR on datasources. That is independent, so if we have to wait with this, I'd give it priority :)

**Comment by PabloAMC** [comment]
Now the backbones are the ones from simple models, we can perhaps implement others (Schnet, DimeNet would be good candidates), but let me know what you think @ethanwharris

**Comment by ethanwharris** [comment]
@PabloAMC Nice! Yeah I think we should wait until torch-geometric releases a new version. If you are able to add additional models by then we can add them here. Otherwise that can always be done in separate PRs :smiley: In the meantime I'll prioritize the data sources work. Thanks!

File: flash/graph/embedding/__init__.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+from flash.graph.classification.data import GraphClassificationData
+from flash.graph.embedding.model import GraphEmbedder
```
Comment (ethanwharris) [inline flash/graph/embedding/__init__.py:2]:
```suggestion
from flash.graph.classification.data import GraphClassificationData  # noqa: E402
from flash.graph.embedding.model import GraphEmbedder  # noqa: E402
```

File: flash_examples/graph_classification.py [hunk 2]
```diff
@@ -22,15 +22,17 @@
 from torch_geometric.datasets import TUDataset  # noqa: E402
 
 # 1. Create the DataModule
-dataset = TUDataset(root="data", name="KKI")
+dataset = TUDataset(root="data", name="KKI").shuffle()
```
Comment (ethanwharris) [inline flash_examples/graph_classification.py:25]: We shuffle before `val_split` anyway so shouldn't need it here :smiley:
```suggestion
dataset = TUDataset(root="data", name="KKI")
```

File: flash_examples/graph_embedding.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    test_dataset=dataset[len(dataset) // 2:],
+    val_split=0.1,
+    train_transform=transform,
+    val_transform=transform,
+    predict_transform=transform,
+)
+"""
+
+# 2. Load a previously trained model (for example, if you have previously run a classification task)
+model = GraphEmbedder.load_from_checkpoint("graph_classification.pt")
```
Comment (ethanwharris) [inline flash_examples/graph_embedding.py:36]: What should be the API here? Presumably this could fail if you try to load from a classifier checkpoint? Should we instead have a method `as_embedder` or something which converts an existing classifier to an ambedder?

File: requirements/datatype_graph.txt [hunk 4]
```diff
@@ -1,3 +1,5 @@
 torch-scatter
 torch-sparse
 torch-geometric
```
Comment (ethanwharris) [inline requirements/datatype_graph.txt:3]: We should set the lates torch geometric as the minimum version here

**Review by ethanwharris — COMMENTED** [review]
Awesome! Looking good :smiley: Some comments

File: flash/graph/backbones.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+# limitations under the License.
+from functools import partial
+
+from flash.core.registry import FlashRegistry
+from flash.core.utilities.imports import _GRAPH_AVAILABLE
+
+if _GRAPH_AVAILABLE:
+    from torch_geometric.nn.models import GAT, GCN, GIN, GraphSAGE
+
+    MODELS = {"GCN": GCN, "GraphSAGE": GraphSAGE, "GAT": GAT, "GIN": GIN}
```
Comment (tchaton) [inline flash/graph/backbones.py:22]: Do you think we could integrate GraphGym directly there ?

File: flash/graph/embedding/model.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    @classmethod
+    def load_from_checkpoint(
+        cls,
+        checkpoint_path: Union[str, IO],
+        map_location: Optional[Union[Dict[str, str], str, torch.device, int, Callable]] = None,
+        hparams_file: Optional[str] = None,
+        strict: bool = True,
+        **kwargs,
+    ) -> "GraphEmbedder":
+        classifier = GraphClassifier.load_from_checkpoint(
```
Comment (tchaton) [inline flash/graph/embedding/model.py:77]: Nice hack :)

**Review by tchaton — APPROVED** [review]
LGTM !

File: flash/graph/classification/model.py [hunk 7]
```diff
@@ -176,3 +102,25 @@ def test_step(self, batch: Any, batch_idx: int) -> Any:
 
     def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Any:
         return super().predict_step(batch, batch_idx, dataloader_idx=dataloader_idx)
+
+    def forward(self, data) -> torch.Tensor:
+        x = self.backbone(data.x, data.edge_index)
+        x = global_mean_pool(x, data.batch)
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/graph/classification/model.py]: Done :smiley:

File: flash/graph/backbones.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from functools import partial
+
+from flash.core.registry import FlashRegistry
+from flash.core.utilities.imports import _GRAPH_AVAILABLE
+from flash.core.utilities.providers import _PYTORCH_GEOMETRIC
+
+if _GRAPH_AVAILABLE:
+    from torch_geometric.nn.models import GAT, GCN, GIN, GraphSAGE
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/graph/backbones.py:21]: I think it's not possible to do cleanly at the moment as the models package also contains many other things that wouldn't work in the GraphClassifier
